Angular: Creating an item component: Minor issues

tes-balo opened the pull request:
Fixes #12675

- **Docs**: Correct statement under **Add HTML for the ItemComponent** heading 

1. "The double curly braces, {{}}, in the <input> and <label> for the checkbox 
signifies Angular's interpolation" is replaced with **The double curly braces, 
\{{}}, in the <label> for the checkbox 
signifies Angular's interpolation**

2.  Refactor editable property in HTML for better readability and less confusion 
by setting `editable` to `true` in `ItemComponent`class
3.  Exchange Boolean values for `editable` where applicable

- **Docs**:  **Prepare the AppComponent** heading
add type parameter to `remove` method in `AppCompponent` class

- **Docs**: Modify logic in `ItemComponent` 

1. `saveItem` method parameter has no type. Set type to `Item` Interface. 
2. `@input() item: Item` has no initializer. Add non-null assertion operator `(!)` to fix 
3. `@input() newItem: ` has no initializer. Add non-null assertion operator `(!)` to fix 
5. `editable`: set editable property to true in **Item Component class**. Setting editable to
 false in the class may lead to confusion in the HTML code
> **Note**: Application will not compile without applying **1, 2 and 3**


- **Docs**: Use the `ItemComponent` in the `AppComponent`
- Replace `item` variable with `i`
- This is to avoid confusing the reader
There's already an `@input` variable named `item` in the `Item Component`, 
using `i` let's the reader know that we are specifically referencing  
members of the `items` array,  and not the `item` Input

- **Fix**: fix typo
typo has already been fixed and merged in issue #20473

<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->
- Changes aim to ensure better readability of tutorial for readers
- Changes fix app compilation errors


### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->
- I just completed the Angular tutorial and I experienced some of the problems listed above. 
- My goal is to simplify the learning experience process for newcomers

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests
Fixes #12675
Relates to #20473 

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

tes-balo committed:
Creating an item component: Couple minor issues

Fixes #12675

- docs: Correct statement under **Add HTML for the ItemComponent** heading

"The double curly braces, {{}}, in the <input> and <label> for the checkbox signifies Angular's interpolation" is replaced with
**The double curly braces, \{{}}, in the <label> for the checkbox signifies Angular's interpolation**
Refactor editable property in HTML for better readability and less confusion by setting editable to true in class
Exchange Boolean values for editable where applicable

docs:  **Prepare the AppComponent** heading
add type parameter to `remove` method in AppCompponent class

docs: Add logic to ItemComponent 

saveItem method parameter has no type. Set ttype to Item Interface.
@input() item: Item has no initializer. Add non-null assertion operator (!) to fix
@input() newItem: newItem has no initializer. Add non-null assertion operator (!) to fix
editable: set editable property to true in Item Component class
setting editable to false in the class may lead to confusion in the HTML code


docs: Use the ItemComponent in the AppComponent

Replace `item` variable with `i`
This is to avoid confusing the reader
There's already an input variable named item in the Item Component
Using i let's the reader know that we are specifically referencing the members of the `items` array, 
and not the `item` Input

fix: fix typo
typo has already been fixed and merged from issue #20473

tes-balo committed:
Merge branch 'main' into patch-1

Josh-Cena commented on files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/angular_item_component/index.md:
```diff
@@ -206,17 +206,19 @@ When you use a property in the template, you must also declare it in the class.
 An `@Input()` serves as a doorway for data to come into the component, and an `@Output()` acts as a doorway for data to go out of the component.
 An `@Output()` has to be of type `EventEmitter`, so that a component can raise an event when there's data ready to share with another component.
 
+> **Note**: The `!` in the class is called the non-null assertion operator. This operator tells Typescript that the `item` and `newItem` Inputs are never `null` or `undefined`. If this operator is not included in your code, the app will fail to compile.
```
This is not the non-null assertion. It's the definite assignment operator. This is non-null assertion:

```js
const a = b! + 1;
```

slim-codes replied:
Thanks for pointing that out @Josh-Cena. 
I had no knowledge of the **definite assignment operator** until now.

tes-balo committed:
Update files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/angular_item_component/index.md

docs: fix incorrect statement

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

Josh-Cena commented on files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/angular_item_component/index.md:
```diff
@@ -83,13 +83,13 @@ Add markup for managing items by replacing the placeholder content in `item.comp
 
   <!-- This section shows only if user clicks Edit button -->
-  <div *ngIf="editable">
+  <div *ngIf="!editable">
```
You seem to have reversed all `editable`. Why? My understanding is that `editable = true` means the model is "editable" and should therefore show the input.

slim-codes replied:
I simply refactored the code. The manner in which `editable` was presented is what I changed. Note that `editable` was previously set to `false` in the class. Then it was negated in the HTML (`*ngIf="!editable"`). `!editable` in that case would still evaluate to `true`.
Reading through the HTML for the first time, one would assume that `!editable` is `false` in the first div which it isn't. That's the main reason why I reversed all of them. Although, you make a good point. I think it would be better to include something like:

> **Note**: `editable = false` in the class, therefore, `!editable` in the first div evaluates to true.

below **line 121** with `editable` in the **non-reversed** form of course. What do you think?

Josh-Cena replied:
The problem is, `editable` is a variable with clear-defined semantics: if it's `true`, the component should show the input elements. Is that the case after your refactor?

slim-codes replied:
No it is not.
Sorry for the delayed response, I've been under the weather lately.
I stand corrected on the refactoring of the `editable` property.
Apparently, I did not fully understand the rationale behind the 
original implementation of the property. 
Now I do

tes-balo committed:
undo changes to the editable property

revert: revert changes to the editable property

changes are unnecessary since editable is a
clear-defined property and as such it should remain as it was

Josh-Cena commented on files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/angular_item_component/index.md:
```diff
@@ -206,10 +206,12 @@ When you use a property in the template, you must also declare it in the class.
 An `@Input()` serves as a doorway for data to come into the component, and an `@Output()` acts as a doorway for data to go out of the component.
 An `@Output()` has to be of type `EventEmitter`, so that a component can raise an event when there's data ready to share with another component.
 
+> **Note**: The `!` in the class's property declaration is called a [definite assignment assertion](https://www.typescriptlang.org/docs/handbook/2/classes.html#--strictpropertyinitialization). This operator tells Typescript that the `item` and `newItem` inputs are always initialized and not `undefined`, even when TypeScript cannot tell from the constructor's definition. If this operator is not in ...
```
```suggestion
> **Note**: The `!` in the class's property declaration is called a [definite assignment assertion](https://www.typescriptlang.org/docs/handbook/2/classes.html#--strictpropertyinitialization). This operator tells Typescript that the `item` and `newItem` inputs are always initialized and not `undefined`, even when TypeScript cannot tell from the constructor's definition. If this operator is not included in your code and you have strict TypeScript compilation settings, the app will fail to compile.
```

Josh-Cena committed:
Update files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/angular_item_component/index.md

Josh-Cena reviewed (approved):
Thank you!
